pkg/remote/gitlab/gitlab.go

```diff
@@ -0,0 +1,374 @@
+	var path = ns(repo.Owner, repo.Name)
+	return client.RepoRawFile(path, build.Commit.Sha, ".drone.yml")
+}
+
+func (r *Gitlab) Status(u *common.User, repo *common.Repo, b *common.Build) error {
+	// how can I get Gitlab status
+	return nil
+}
+
+// Netrc returns a .netrc file that can be used to clone
+// private repositories from a remote system.
+func (r *Gitlab) Netrc(u *common.User) (*common.Netrc, error) {
+	url_, err := url.Parse(r.URL)
+	if err != nil {
+		return nil, err
+	}
+	netrc := &common.Netrc{}
+	netrc.Login = u.Token
+	netrc.Password = "x-oauth-basic"
+	netrc.Machine = url_.Host
+	return netrc, nil
+}
+
+// Activate activates a repository by adding a Post-commit hook and
+// a Public Deploy key, if applicable.
+func (r *Gitlab) Activate(user *common.User, repo *common.Repo, keys *common.Keypair, link string) error {
+	var client = NewClient(r.URL, user.Token, r.SkipVerify)
+	var path = ns(repo.Owner, repo.Name)
+	var title, err = GetKeyTitle(link)
+	if err != nil {
+		return err
+	}
+
+	// if the repository is private we'll need
+	// to upload a github key to the repository
+	if repo.Private {
+		var err = client.AddProjectDeployKey(path, title, repo.Keys.Public)
+		if err != nil {
+			return err
+		}
+	}
+
+	// append the repo owner / name to the hook url since gitlab
+	// doesn't send this detail in the post-commit hook
+	link += "?owner=" + repo.Owner + "&name=" + repo.Name
+
+	// add the hook
+	return client.AddProjectHook(path, link, true, false, true)
+}
+
+// Deactivate removes a repository by removing all the post-commit hooks
+// which are equal to link and removing the SSH deploy key.
+func (r *Gitlab) Deactivate(user *common.User, repo *common.Repo, link string) error {
+	var client = NewClient(r.URL, user.Token, r.SkipVerify)
+	var path = ns(repo.Owner, repo.Name)
+
+	keys, err := client.ProjectDeployKeys(path)
+	if err != nil {
+		return err
+	}
+	var pubkey = strings.TrimSpace(repo.Keys.Public)
+	for _, k := range keys {
+		if pubkey == strings.TrimSpace(k.Key) {
+			if err := client.RemoveProjectDeployKey(path, strconv.Itoa(k.Id)); err != nil {
+				return err
+			}
+			break
+		}
+	}
+	hooks, err := client.ProjectHooks(path)
+	if err != nil {
+		return err
+	}
+	link += "?owner=" + repo.Owner + "&name=" + repo.Name
+	for _, h := range hooks {
+		if link == h.Url {
+			if err := client.RemoveProjectHook(path, strconv.Itoa(h.Id)); err != nil {
+				return err
+			}
+			break
+		}
+	}
+	return nil
+}
+
+// ParseHook parses the post-commit hook from the Request body
+// and returns the required data in a standard format.
+func (r *Gitlab) Hook(req *http.Request) (*common.Hook, error) {
+
+	defer req.Body.Close()
+	var payload, _ = ioutil.ReadAll(req.Body)
+	var parsed, err = gogitlab.ParseHook(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(parsed.After) == 0 || parsed.TotalCommitsCount == 0 {
+		return nil, nil
+	}
+
+	if parsed.ObjectKind == "merge_request" {
```

Comment:
In drone 0.4 we store source, know about source and target projects https://github.com/drone/drone/pull/634/files